File: src/main/java/br/com/alura/forum/service/DashboardDataProcessingService.java
package br.com.alura.forum.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import br.com.alura.forum.model.Category;
import br.com.alura.forum.repository.CategoryRepository;
import br.com.alura.forum.vo.CategoriesAndTheirStatisticsData;
import br.com.alura.forum.vo.CategoryStatisticsData;

@Service
public class DashboardDataProcessingService {

	private CategoryRepository categoryRepository;	
	private CategoryStatisticsDataLoadingService categoryStatisticsService;

	@Autowired
	public DashboardDataProcessingService(CategoryRepository categoryRepository, 
			CategoryStatisticsDataLoadingService categoryStatisticsService) {
				
		this.categoryRepository = categoryRepository;
		this.categoryStatisticsService = categoryStatisticsService;
	}
	
	public CategoriesAndTheirStatisticsData execute() {
		List<Category> principalCategories = this.categoryRepository.findByCategoryIsNull();
		
		CategoriesAndTheirStatisticsData categoriesAndTheirData = new CategoriesAndTheirStatisticsData();

		principalCategories.stream().forEach(category -> {
			CategoryStatisticsData statisticsData = this.categoryStatisticsService.load(category);
			categoriesAndTheirData.add(category, statisticsData);
		});
				
		return categoriesAndTheirData;
	}
}

File: src/main/kotlin/br/com/alura/forum/service/DashboardDataProcessingService.kt
package br.com.alura.forum.service

import br.com.alura.forum.repository.CategoryRepository
import br.com.alura.forum.vo.CategoriesAndTheirStatisticsData
import org.springframework.stereotype.Service
import java.time.Instant
import java.time.temporal.ChronoUnit

@Service
class DashboardDataProcessingService(
        private val categoryRepository: CategoryRepository,
        private val categoryStatisticsService: CategoryStatisticsDataLoadingService
) {

    fun execute(lastWeek: Instant): CategoriesAndTheirStatisticsData {
        val categoriesAndTheirData = CategoriesAndTheirStatisticsData()

        val principalCategories = categoryRepository.findByCategoryIsNull()

        principalCategories.forEach {
            categoriesAndTheirData.add(it, categoryStatisticsService.load(it, lastWeek))
        }

        return categoriesAndTheirData
    }
}